### PR TITLE
Fixing issues uncovered by memcheck [complex-fem-memcheck-dev]

### PR DESCRIPTION
**Pull request by mlstowell**
After adding the new examples 19 - 22 to the `memcheck` tests a few issues appeared.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/1234 | @mlstowell | @v-dobrev | @v-dobrev + @tzanio | 01/16/20 | 01/16/20 | ⌛due 01/23/20 |

### CODE REVIEW

**Comment by mlstowell** [comment]
Hi, @v-dobrev ,
You may want to look at an issue that appeared with the prolongation operator https://github.com/mfem/mfem/blob/0cfffe7ff7bed4028334b26219ba1fc1655ee507/fem/pfespace.cpp#L879
You recently modified this condition statement to resolve an issue with the `OperatorJacobiSmoother`.  Unfortunately, this leads to a problem with `FormLinearSystem` when the number of processors is equal to 1:
https://github.com/mfem/mfem/blob/0cfffe7ff7bed4028334b26219ba1fc1655ee507/fem/pbilinearform.cpp#L304
In this case the pointer retuned to `FormLinearSystem` can be NULL.

Best wishes,
Mark

File: examples/ex19p.cpp [hunk 1]
```diff
@@ -348,8 +348,6 @@ int main(int argc, char *argv[])
    delete pmesh;
 
    MPI_Finalize();
-
-   return 0;
```
Comment (tzanio) [inline examples/ex19p.cpp]: Why not?

Comment (mlstowell) [reply examples/ex19p.cpp]: I cannot seem to reproduce the behavior I saw last night.  I'll put these back and run more tests.

**Comment by v-dobrev** [comment]
> Hi, @v-dobrev ,
> You may want to look at an issue that appeared with the prolongation operator
> 
> https://github.com/mfem/mfem/blob/0cfffe7ff7bed4028334b26219ba1fc1655ee507/fem/pfespace.cpp#L879
> 
> You recently modified this condition statement to resolve an issue with the `OperatorJacobiSmoother`. Unfortunately, this leads to a problem with `FormLinearSystem` when the number of processors is equal to 1:
> 
> https://github.com/mfem/mfem/blob/0cfffe7ff7bed4028334b26219ba1fc1655ee507/fem/pbilinearform.cpp#L304
> 
> In this case the pointer retuned to `FormLinearSystem` can be NULL.
> 
> Best wishes,
> Mark

Yes, this was intentional -- to unify the behavior (on 1 MPI rank) on CPU and GPU and also draw attention to this issue. We should discuss how we want to address it and fix it soon.

**Comment by v-dobrev** [comment]
By the way, the valgrind tests also show memory leaks in `ex19p` -- at first glance, these maybe in hypre but we should double-check.

**Comment by mlstowell** [comment]
Hi, @v-dobrev ,
The memory leak in `ex19p` occurs during setup of HYPRE's AMG solver in function `hypre_BoomerAMG_GMExpandInterp`.  This particular leak was fixed in November 2017, according to GitHub, and probably appears in HYPRE's 2.14.0 version.
Mark

**Comment by v-dobrev** [comment]
Hmm, OK. Then we should probably disable the testing of `ex19p` in our `memcheck` script, at least until we update the hypre version we use there.

**Comment by tzanio** [comment]
Another option could be to add this as an exception in https://mybitbucket.llnl.gov/projects/MFEM/repos/tests/browse/memcheck.supp

**Pull request merged**